### PR TITLE
Adds support for rgba and rgb images to pcolorfast

### PR DESCRIPTION
**Pull request by paalge**
## PR Summary

This pull request fixes issue #1317, by making nr, and nc equal to the first and second element of C.shape.
Through this, pcolorfast now works on RGBA arrays, i.e.  NxMx3 or NxMx4.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/whats_new.rst if major new feature
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by efiring** [comment]
Please look again at the discussion in #1317.  This simple 1-line change, by itself, is not enough.

**Comment by paalge** [comment]
I have read the discussion, and I agree the one line is not enough, though it fixes half of the problem, instead of none of it as it has been for the last years, while generating an exception (as it does now) for some complicated grids. This PR fixes problems I had with the code, and would for others as well.

**Comment by dopplershift** [comment]
Reading the discussion, it seems like there wasn't anything against making this particular change, though.

**Comment by efiring** [comment]
Correct.  It just needs a complete PR, with more than the one-line change.

1) Additions to the docstring explaining the new behavior, and when it is available.
2) Validation with an informative error message, to trap the case where a 3-D array makes its way into the quadmesh code path.
3) A note in doc/api/api_changes.

**Comment by dopplershift** [comment]
Thank you @efiring for being explicit in what changes you were requesting... 😁

**Comment by paalge** [comment]
Thank you @efiring for specifying what you wanted.
I've pushed some changes now that should address these.

**Comment by paalge** [comment]
I see this was not part of the v/2.2 release. What is needed to get this into matplotlib?

**Comment by anntzer** [comment]
Needs a rebase, but I think something like this can go in?  Possibly with some tests as well...

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -5602,6 +5602,10 @@ def pcolorfast(self, *args, **kwargs):
         C is the 2D array of color values corresponding to quadrilateral
         cells. Let (nr, nc) be its shape.  C may be a masked array.
 
+        C can in some cases be 3D with the last dimension as rgb(a).
+        This is available when C qualifies for image or pcolorimage type,
+        will throw a TyepError if C is 3D and quadmesh
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: -> "otherwise if C is 3D a TypeError will be raised."

File: doc/api/api_changes/2017-06-08-PGE.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Added support for RGB(A) images in pcolorfast
+`````````````````````````````````````````````
+
+pcolorfast now excepts 3D images (x,y,color) as inputs in `C` if they qualify for pcolorimage or image
```
Comment (efiring) [inline doc/api/api_changes/2017-06-08-PGE.rst]: "excepts..." -> "accepts 3D (RGB or RGBA) arrays if the X and Y specifications allow image or pcolorimage rendering; they remain unsupported by the more general quadmesh rendering."

**Comment by paalge** [comment]
I've added a test, minor doc change and rebased.
I hope it is okay now?

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -5230,6 +5230,26 @@ def test_pcolor_fast_non_uniform():
...
+    C = np.random.rand(10, 10, 3)  # RGB image [0,1]
+    x = np.arange(11, dtype=np.float)
+    y = np.arange(11, dtype=np.float)
+
+    ax.pcolorfast(C)
+    ax.pcolorfast(x, y, C)
+
+    xv, yv = np.meshgrid(x, y)
+
+    try:
```
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: use pytest.raises

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -6211,6 +6218,9 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
                     else:
                         style = "pcolorimage"
             elif x.ndim == 2 and y.ndim == 2:
+                if C.ndim > 2:
+                    raise TypeError('''Pcolorfast needs to use quadmesh, but C has more then 2 dimensions.
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: break the too-long line.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -6104,8 +6104,15 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
 
         .. warning::
 
-           This method is experimental. Compared to `~.Axes.pcolor` or
-           `~.Axes.pcolormesh` it has some limitations:
+        C can in some cases be 3D with the last dimension as rgb(a).
+        This is available when C qualifies for image or pcolorimage type,
+        otherwise if C is 3D, a TypeError will be raised.
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: I think this should raise a ValueError instead (the types are all correct).

File: lib/matplotlib/tests/test_axes.py (hunk 3 above)
Comment (paalge) [reply lib/matplotlib/tests/test_axes.py]: Done

File: lib/matplotlib/axes/_axes.py (hunk 4 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: Done

File: lib/matplotlib/axes/_axes.py (hunk 5 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: I guess that might be better. If others agree I can change this.

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -6211,6 +6218,10 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
                     else:
                         style = "pcolorimage"
             elif x.ndim == 2 and y.ndim == 2:
+                if C.ndim > 2:
+                    raise TypeError('''
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: Yes ValueError.  This message doesn't actually make much sense to me; I don't think a naive user has any idea what Quadmesh is, nor will they care.  Can this error message just say what the error is and how to fix it without asking any questions...

File: lib/matplotlib/axes/_axes.py (hunk 5 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: Changed

File: lib/matplotlib/axes/_axes.py (hunk 6 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: I've changed the error type and tried to write a better error message

File: lib/matplotlib/tests/test_axes.py [hunk 7]
```diff
@@ -4669,7 +4669,7 @@ def test_rcparam_grid_minor():
         (('both'), (True, True)),
         (('major'), (True, False)),
         (('minor'), (False, True))
-        )
+    )
```
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: please undo this change

File: lib/matplotlib/tests/test_axes.py [hunk 8]
```diff
@@ -5230,6 +5230,24 @@ def test_pcolor_fast_non_uniform():
...
+
+    fig, ax = plt.subplots(1, 1)
+
+    np.random.seed(19680801)
+    C = np.random.rand(10, 10, 3)  # RGB image [0,1]
+    x = np.arange(11, dtype=np.float)
+    y = np.arange(11, dtype=np.float)
+
+    ax.pcolorfast(C)
+    ax.pcolorfast(x, y, C)
```
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: The test on this line and the one just above are now taken care of by test_pcolorfast_colormapped, so they can be removed.

File: lib/matplotlib/axes/_axes.py [hunk 9]
```diff
@@ -6177,9 +6177,15 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
 
         .. warning::
+           C can in some cases be 3D with the last dimension as rgb(a).
+           This is available when C qualifies for image or pcolorimage type,
+           otherwise if C is 3D, a TypeError will be raised.
 
-           This method is experimental. Compared to `~.Axes.pcolor` or
-           `~.Axes.pcolormesh` it has some limitations:
+        ``ax.pcolorfast(C, **kwargs)`` is equivalent to
+        ``ax.pcolorfast([0,nc], [0,nr], C, **kwargs)``
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: These two lines add nothing to the docstring that is not available in the docs for X and Y, so should be removed.

File: lib/matplotlib/axes/_axes.py [hunk 10]
```diff
@@ -6177,9 +6177,15 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
+
+        This method is experimental. Compared to `~.Axes.pcolor` or
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: keep these two lines indented (basically, undo the change)

**Comment by anntzer** [comment]
Also, consider perhaps squashing your commits.

File: lib/matplotlib/axes/_axes.py [hunk 11]
```diff
@@ -6285,6 +6291,10 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
                     else:
                         style = "pcolorimage"
             elif x.ndim == 2 and y.ndim == 2:
+                if C.ndim > 2:
+                    raise ValueError('''
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: see e.g. https://github.com/matplotlib/matplotlib/blob/17d16f972f43d3cec538bc27f8006ba1913dbc18/lib/matplotlib/axes/_axes.py#L5663-L5666 for how to format a long error message.

**Comment by anntzer** [comment]
I would like something like this to make it to 3.1, especially considering that it's quite close to being mergeable.  Feel free to remilestone if you disagree.

File: lib/matplotlib/tests/test_axes.py (hunk 7 above)
Comment (paalge) [reply lib/matplotlib/tests/test_axes.py]: Done in https://github.com/matplotlib/matplotlib/pull/8690/commits/af603a93495230161eb3c2354236f986b15769f8

File: lib/matplotlib/axes/_axes.py (hunk 9 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: Fixed in https://github.com/matplotlib/matplotlib/pull/8690/commits/af603a93495230161eb3c2354236f986b15769f8

File: lib/matplotlib/tests/test_axes.py (hunk 8 above)
Comment (paalge) [reply lib/matplotlib/tests/test_axes.py]: Okay, they have been removed https://github.com/matplotlib/matplotlib/pull/8690/commits/af603a93495230161eb3c2354236f986b15769f8

File: lib/matplotlib/axes/_axes.py (hunk 10 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: Fixed in https://github.com/matplotlib/matplotlib/pull/8690/commits/af603a93495230161eb3c2354236f986b15769f8

File: lib/matplotlib/axes/_axes.py (hunk 11 above)
Comment (paalge) [reply lib/matplotlib/axes/_axes.py]: Fixed in https://github.com/matplotlib/matplotlib/pull/8690/commits/af603a93495230161eb3c2354236f986b15769f8

**Comment by paalge** [comment]
I've addressed the comments from @anntzer. 
I'll squash the commits if there are no other comments.

File: lib/matplotlib/axes/_axes.py [hunk 12]
```diff
@@ -6177,9 +6177,12 @@ def pcolorfast(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
 
         .. warning::
+           C can in some cases be 3D with the last dimension as rgb(a).
+           This is available when C qualifies for image or pcolorimage type,
+           otherwise if C is 3D, a TypeError will be raised.
 
-           This method is experimental. Compared to `~.Axes.pcolor` or
-           `~.Axes.pcolormesh` it has some limitations:
+         This method is experimental. Compared to `~.Axes.pcolor` or
+         `~.Axes.pcolormesh` it has some limitations:
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: again, there should be no diff at these two lines

**Comment by anntzer** [comment]
Only one minor comment left, you can fix and squash.

**Comment by jklymak** [comment]
Feel free to re-milestone if this gets done sooner, but I don't think its urgent...

**Comment by paalge** [comment]
I've changed the last comment, but I having problems squashing the pull request due to the rebases (4300 commands that git needs to do). This could be due to my inexperience in squashing. If it can't be done in when you merge the code into master, my solution would be to close this and move the code to a new pull request based on the latest master

**Comment by anntzer** [comment]
Do you mind if I force-push to your branch to fix that?

**Comment by paalge** [comment]
@anntzer Go ahead, you're more than welcome.

**Comment by anntzer** [comment]
Rebased.

**Comment by anntzer** [comment]
Good to go for 3.1.0 IMO.
